packages/runtime/src/next/components/page.tsx

```diff
@@ -1,51 +1,68 @@
 'use client'
 
-import { Suspense, memo, useMemo } from 'react'
+import { Suspense, memo, useEffect } from 'react'
 
-import { RuntimeProvider } from '../../runtimes/react'
-import { Page as PageMeta } from '../../components/page'
-import { MakeswiftHostApiClient } from '../../api/react'
-import { MakeswiftPageSnapshot } from '../client'
+import { type CacheData } from '../../api/react'
+
+import { useDispatch } from '../../runtimes/react/hooks/use-dispatch'
+import { useMakeswiftHostApiClient } from '../../runtimes/react/host-api-client'
+
+import { Page as PageComponent } from '../../components/page'
+import {
+  type MakeswiftPageSnapshot,
+  type MakeswiftPageDocument,
+  pageToRootDocument,
+} from '../client'
+
+import * as ReactPage from '../../state/react-page'
+import { registerDocumentsEffect, updateAPIClientCache } from '../../state/actions'
 
 export type PageProps = {
   snapshot: MakeswiftPageSnapshot
 }
 
+const isServer = typeof window === 'undefined'
+
+function useCacheData(cacheData: CacheData) {
+  const client = useMakeswiftHostApiClient()
+
+  if (isServer) {
+    client.makeswiftApiClient.dispatch(updateAPIClientCache(cacheData))
+  }
+
+  useEffect(() => {
+    client.makeswiftApiClient.dispatch(updateAPIClientCache(cacheData))
+  }, [cacheData, client])
+}
+
+function useRegisterPageDocument(pageDocument: MakeswiftPageDocument): ReactPage.Document {
+  const dispatch = useDispatch()
+
+  const rootDocument = pageToRootDocument(pageDocument)
+  useEffect(() => dispatch(registerDocumentsEffect([rootDocument])), [rootDocument])
+  return rootDocument
+}
```

Comment:
@migueloller Is this more along the lines of what you envisioned re: moving application logic out of React?

Reply:
Yes, for sure. In this case, though, I'd move the `pageToRootDocument` inside the effect and make the dependency `pageDocument` so that `pageToRootDocument` doesn't have to be memoized or attempt to preserve the reference equality of `pageDocument`. You could alternatively use `useMemo`.